File: ios/Classes/GiphyFlutterGridView.swift
//
//  GiphyFlutterGridView.swift
//  giphy_flutter_sdk
//
//  Created by Alex on 4.3.2024.
//

import Flutter
import UIKit
import GiphyUISDK

public class GiphyFlutterGridView: NSObject, FlutterPlatformView {
    
    private let gridController: GiphyGridController = GiphyGridController()
    
    private var channel: FlutterMethodChannel?
    
    init(frame: CGRect, viewId: Int64, args: Any?, binaryMessenger messenger: FlutterBinaryMessenger) {
        super.init()
        gridController.delegate = self
        setupView(frame: frame)
        createMethodChannel(viewId: viewId, messenger: messenger)
    }
    
    deinit {
        channel?.setMethodCallHandler(nil)
    }
    
    public func view() -> UIView {
        return gridController.view
    }
    
    private func setContent(_ value: [String: Any?]) -> Void {
        let rating = GPHRatingType.fromString(value["rating"] as? String)
        gridController.rating = rating ?? GPHRatingType.ratedPG13
        gridController.content = GPHContent.fromDictionary(value)
        gridController.update()
    }
    
    private func setCellPadding(_ value: CGFloat) -> Void {
        gridController.cellPadding = value
    }
    
    private func setClipsPreviewRenditionType(_ value: String?) -> Void {
        let renditionType = GPHRenditionType.fromString(value)
        gridController.clipsPreviewRenditionType = renditionType ?? .fixedWidth
    }
    
    private func setDisableEmojiVariations(_ value: Bool) -> Void {
        gridController.emojiSettings.disableVariations = value
    }
    
    private func setRenditionType(_ value: String?) -> Void {
        let renditionType = GPHRenditionType.fromString(value)
        gridController.renditionType = renditionType ?? .fixedWidth
    }
    
    private func setOrientation(_ value: String?) -> Void {
        gridController.direction = UICollectionView.ScrollDirection.fromString(value ?? "")
    }
    
    private func setSpanCount(_ value: Int) -> Void {
        gridController.numberOfTracks = value
    }
    
    private func setFixedSizeCells(_ value: Bool) -> Void {
        gridController.fixedSizeCells = value
    }
    
    private func setTheme(_ value: [String: Any?]) -> Void {
        gridController.theme = GiphyFlutterTheme(flutterConfig: value)
    }
    
    private func createMethodChannel(viewId: Int64, messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: "com.giphyfluttersdk/gridView\(viewId)", binaryMessenger: messenger)
        channel?.setMethodCallHandler({
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            guard let self = self else {
                result(FlutterError(code: "SELF_NIL", message: "Reference to self is nil", details: nil))
                return
            }
            
            let arguments = call.arguments as? [String: Any?] ?? [:]
            
            switch call.method {
            case "setContent":
                if let value = arguments["content"] as? [String: Any?] {
                    self.setContent(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a map for content", details: nil))
                }
            case "setCellPadding":
                if let value = arguments["cellPadding"] as? CGFloat {
                    self.setCellPadding(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a Float for cellPadding", details: nil))
                }
            case "setDisableEmojiVariations":
                if let value = arguments["disableEmojiVariations"] as? Bool {
                    self.setDisableEmojiVariations(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a Bool for disableEmojiVariations", details: nil))
                }
            case "setFixedSizeCells":
                if let value = arguments["fixedSizeCells"] as? Bool {
                    self.setDisableEmojiVariations(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a Bool for fixedSizeCells", details: nil))
                }
            case "setOrientation":
                if let value = arguments["orientation"] as? String {
                    self.setOrientation(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a String for orientation", details: nil))
                }
            case "setSpanCount":
                if let value = arguments["spanCount"] as? Int {
                    self.setSpanCount(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected an Int for orientation", details: nil))
                }
            case "setRenditionType":
                if let value = arguments["renditionType"] as? String {
                    self.setRenditionType(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a String for renditionType", details: nil))
                }
            case "setClipsPreviewRenditionType":
                if let value = arguments["renditionType"] as? String {
                    self.setClipsPreviewRenditionType(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a String for renditionType", details: nil))
                }
            case "setShowCheckeredBackground":
                // To maintain consistency with Android, do not return an error, even though it's not supported.
                result(nil)
            case "setTheme":
                if let value = arguments["theme"] as? [String: Any?] {
                    self.setTheme(value)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a map for theme", details: nil))
                }
            default:
                result(FlutterMethodNotImplemented)
            }
        })
    }
    
    private func setupView(frame: CGRect) -> Void {
        gridController.view.frame = frame
        gridController.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    }
}

extension GiphyFlutterGridView: GPHGridDelegate {
    public func contentDidUpdate(resultCount: Int, error: (Error)?) {
        channel?.invokeMethod("onContentUpdate", arguments: ["resultCount": resultCount])
    }
    
    public func didSelectMedia(media: GiphyUISDK.GPHMedia, cell: UICollectionViewCell) {
        channel?.invokeMethod("onMediaSelect", arguments: [
            "media": media.toDictionary()
        ])
    }
    
    public func didSelectMoreByYou(query: String) {
        // pass
    }
    
    public func didScroll(offset: CGFloat) {
        channel?.invokeMethod("onScroll", arguments: [
            "offset": offset
        ])
    }
}

File: android/src/main/kotlin/com/giphy/giphy_flutter_sdk/GiphyFlutterGridView.kt
package com.giphy.giphy_flutter_sdk

import android.content.Context
import android.view.View
import com.giphy.giphy_flutter_sdk.dto.GiphyFlutterTheme
import com.giphy.giphy_flutter_sdk.dto.toHashMap
import com.giphy.giphy_flutter_sdk.dto.toRating
import com.giphy.giphy_flutter_sdk.dto.toRenditionType
import com.giphy.giphy_flutter_sdk.utils.require
import com.giphy.sdk.core.models.Media
import com.giphy.sdk.core.models.enums.MediaType
import com.giphy.sdk.core.models.enums.RatingType
import com.giphy.sdk.core.models.enums.RenditionType
import com.giphy.sdk.ui.GPHRequestType
import com.giphy.sdk.ui.pagination.GPHContent
import com.giphy.sdk.ui.views.GPHGridCallback
import com.giphy.sdk.ui.views.GPHSearchGridCallback
import com.giphy.sdk.ui.views.GifView
import com.giphy.sdk.ui.views.GiphyGridView
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView

internal class GiphyFlutterGridView(
    context: Context?,
    messenger: BinaryMessenger,
    viewId: Int,
    params: Map<String, Any>?
) : PlatformView {
    private val view: GiphyGridView
    private val channel = MethodChannel(messenger, "com.giphyfluttersdk/gridView$viewId")
    
    init {
        view = GiphyGridView(context.require("Context must not be null"))
        view.callback = object : GPHGridCallback {
            override fun contentDidUpdate(resultCount: Int) {
                channel.invokeMethod(
                    "onContentUpdate", mapOf(
                        "resultCount" to resultCount
                    )
                )
            }

            override fun didSelectMedia(media: Media) {
                channel.invokeMethod(
                    "onMediaSelect", mapOf(
                        "media" to media.toHashMap()
                    )
                )
            }
        }

        view.searchCallback = object : GPHSearchGridCallback {
            override fun didTapUsername(username: String) {
                // pass
            }

            override fun didLongPressCell(cell: GifView) {
                // pass
            }

            override fun didScroll(dx: Int, dy: Int) {
                channel.invokeMethod(
                    "onScroll", mapOf(
                        "offset" to dy.toDouble()
                    )
                )
            }
        }

        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "setContent" -> {
                    call.argument<HashMap<String, Any?>>("content")?.let {
                        setContent(it)
                    }
                    result.success(null)
                }

                "setCellPadding" -> {
                    setCellPadding(call.argument<Double>("cellPadding"))
                    result.success(null)
                }

                "setDisableEmojiVariations" -> {
                    setDisableEmojiVariations(call.argument<Boolean>("disableEmojiVariations"))
                    result.success(null)
                }

                "setFixedSizeCells" -> {
                    setFixedSizeCells(call.argument<Boolean>("fixedSizeCells"))
                    result.success(null)
                }

                "setOrientation" -> {
                    setOrientation(call.argument<String>("orientation"))
                    result.success(null)
                }

                "setSpanCount" -> {
                    setSpanCount(call.argument<Int>("spanCount"))
                    result.success(null)
                }

                "setRenditionType" -> {
                    setRenditionType(call.arguments as? HashMap<String, Any?>)
                    result.success(null)
                }

                "setClipsPreviewRenditionType" -> {
                    setClipsPreviewRenditionType(call.arguments as? HashMap<String, Any?>)
                    result.success(null)
                }

                "setShowCheckeredBackground" -> {
                    setShowCheckeredBackground(call.argument<Boolean>("showCheckeredBackground"))
                    result.success(null)
                }

                "setTheme" -> {
                    setTheme(call.arguments as? HashMap<String, Any?>)
                    result.success(null)
                }

                else -> result.success(null)//result.notImplemented()
            }
        }
    }

    private fun setContent(value: HashMap<String, Any?>) {
        val content = GPHContent()

        val requestType = (value["requestType"] as? String)
        if (requestType != null) {
            content.requestType = GPHRequestType.values().firstOrNull { it.name == requestType }
                ?: GPHRequestType.search

            if (content.requestType == GPHRequestType.recents) {
                view.content = GPHContent.recents
                return
            }
        }
        content.rating = value.toRating() ?: RatingType.pg13

        (value["searchQuery"] as? String)?.let {
            content.searchQuery = it
        }

        (value["mediaType"] as? String)?.let {
            content.mediaType = MediaType.valueOf(it)
        }

        view.content = content
    }

    private fun setCellPadding(value: Double?) {
        if (value != null) {
            view.cellPadding = value.toInt()
        }
    }

    private fun setDisableEmojiVariations(value: Boolean?) {
        if (value != null) {
            view.disableEmojiVariations = value
        }
    }

    private fun setFixedSizeCells(value: Boolean?) {
        if (value != null) {
            view.fixedSizeCells = value
        }
    }

    private fun setOrientation(value: String?) {
        view.direction = when (value) {
            "horizontal" -> GiphyGridView.HORIZONTAL
            "vertical" -> GiphyGridView.VERTICAL
            else -> GiphyGridView.VERTICAL
        }
    }

    private fun setSpanCount(value: Int?) {
        if (value != null) {
            view.spanCount = value
        }
    }

    private fun setRenditionType(map: HashMap<String, Any?>?) {
        view.renditionType = map?.toRenditionType() ?: RenditionType.fixedWidth
    }

    private fun setClipsPreviewRenditionType(map: HashMap<String, Any?>?) {
        view.clipsPreviewRenditionType = map?.toRenditionType() ?: RenditionType.fixedWidth
    }

    private fun setShowCheckeredBackground(value: Boolean?) {
        view.showCheckeredBackground = value ?: false
    }

    private fun setTheme(map: HashMap<String, Any?>?) {
        val value = map ?: return
        view.theme =
            GiphyFlutterTheme.getInstance().toGPHTheme(value["theme"] as HashMap<String, Any?>)
    }

    override fun getView(): View {
        return view
    }

    override fun dispose() {
        channel.setMethodCallHandler(null)
        view.callback = null
        view.searchCallback = null
    }

}
